compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +154,400 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
+                                                 getConstant(coeffs[dim]));
+      // Case where `combination` is constant, `value` is not constant.
+      if (auto combinationAttr = dyn_cast<Attribute>(combination)) {
+        int64_t c = cast<IntegerAttr>(combinationAttr).getInt();
+        if (c != 0) {
+          combination = builder.create<arith::AddIOp>(loc, getConstant(c), term)
+                            .getResult();
+        } else {
+          combination = term;
+        }
+      } else {
+        // Case where neither `combination` nor `value` is constant.
+        Value combinationVal = cast<Value>(combination);
+        combination = builder.create<arith::AddIOp>(loc, combinationVal, term)
+                          .getResult();
+      }
     }
-    offsets = subviewOp.getMixedOffsets();
-    auto [stridesI64, baseOffset] =
-        getStridesAndOffset(subviewOp.getSource().getType());
-    if (baseOffset != 0) {
-      auto message = llvm::formatv(
-          "has non-zero base offset {0} that is not supported by the "
-          "current pass: requires testing and possible code changes.",
-          baseOffset);
-      return subviewOp->emitOpError(message);
+  }
+  return combination;
+}
+
+
+/// Update the offsets, sizes, and strides from a collapse shape operation.
+LogicalResult updateFromCollapseShape(memref::CollapseShapeOp collapseOp,
+                                      SmallVector<OpFoldResult> &offsets,
+                                      SmallVector<OpFoldResult> &sizes,
+                                      SmallVector<OpFoldResult> &strides) {
+  auto reassociationIndices = collapseOp.getReassociationIndices();
+  ArrayRef<int64_t> inputShape = collapseOp.getSrcType().getShape();
+  ArrayRef<int64_t> resultShape = collapseOp.getType().getShape();
+  uint64_t resultRank = resultShape.size();
+  MLIRContext *ctx = collapseOp.getContext();
+
+  // Set strides to inner-most stride in each reassocation group.
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    uint64_t index = reassociation.index();
+    uint64_t dim = reassociation.value().back();
+    strides[index] = strides[dim];
+  }
+  strides.resize(resultRank);
+
+  // Set sizes to output shape, and check that all dims are static.
+  sizes.clear();
+  for (int64_t dim : resultShape) {
+    if (dim == ShapedType::kDynamic) {
+      return collapseOp.emitOpError(
+          "has a dynamic shape which is currently unsupported.");
     }
-    strides = getAsIndexOpFoldResult(ctx, stridesI64);
-    operandOp = subviewOp.getSource().getDefiningOp();
-    sizes = subviewOp.getMixedSizes();
-    if (llvm::any_of(sizes, [](OpFoldResult fr) {
-          return !getConstantIntValue(fr).has_value();
-        })) {
-      return subviewOp->emitOpError(
-          " has dynamic shape that is not supported by the target dma op.");
+    sizes.push_back(getAsIndexOpFoldResult(ctx, dim));
+  }
+
+  // Offsets - merge reassocation groups.
```

Comment:
I would put some explanation of how offsets are decided here and other places too.